cpp/oneapi/dal/algo/kmeans/backend/gpu/kernels_fp.hpp

```diff
@@ -0,0 +1,459 @@
+
+#pragma once
+
+#include "oneapi/dal/backend/common.hpp"
+#include "oneapi/dal/backend/primitives/ndarray.hpp"
+#include "oneapi/dal/backend/primitives/distance.hpp"
+#include "oneapi/dal/backend/primitives/sort/sort.hpp"
+
+namespace oneapi::dal::kmeans::backend {
+
+#ifdef ONEDAL_DATA_PARALLEL
+
+namespace bk = dal::backend;
+namespace pr = dal::backend::primitives;
+
+inline std::int64_t get_recommended_sg_size(sycl::queue& queue) {
+    // TODO optimization/dispatching
+    return 16;
+}
+
+inline std::int64_t get_recommended_wg_count(sycl::queue& queue) {
+    // TODO optimization/dispatching
+    return 128;
+}
+
+inline std::int64_t get_scaled_wg_size_per_row(const sycl::queue& queue,
+                                               std::int64_t column_count,
+                                               std::int64_t preffered_wg_size) {
+    const std::int64_t sg_max_size = bk::device_max_sg_size(queue);
+    const std::int64_t row_adjusted_sg_num =
+        column_count / sg_max_size + std::int64_t(column_count % sg_max_size > 0);
+    std::int64_t expected_sg_num = std::min(preffered_wg_size / sg_max_size, row_adjusted_sg_num);
+    if (expected_sg_num < 1)
+        expected_sg_num = 1;
+    return dal::detail::check_mul_overflow(expected_sg_num, sg_max_size);
+}
+
+inline std::int64_t get_gpu_sg_size(sycl::queue& queue) {
+    // TODO optimization/dispatching
+    return 16;
+}
+
+namespace bk = dal::backend;
+namespace pr = dal::backend::primitives;
+
+template <typename T>
+struct select_min_distance {};
+
+template <typename T>
+struct set_indices {};
+
+template <typename T>
+struct find_candidates_kernel {};
+
+template <typename T>
+struct fill_empty_cluster_kernel {};
+
+template <typename T>
+struct centroid_reduction {};
+
+template <typename T>
+struct reset_partial_centroids {};
+
+template <typename T>
+struct centroid_merge {};
+
+template <typename T>
+struct compute_obj_function {};
+
+template <typename Float>
+std::int64_t get_block_size_in_rows(sycl::queue& queue, std::int64_t column_count) {
+    // TODO optimization
+    std::int64_t block_size_in_bytes = bk::device_global_mem_cache_size(queue);
+    return block_size_in_bytes / column_count / sizeof(Float);
+}
+
+template <typename Float>
+std::int64_t get_part_count_for_partial_centroids(sycl::queue& queue,
+                                                  std::int64_t column_count,
+                                                  std::int64_t cluster_count) {
+    // TODO optimization
+    const std::int64_t block_size_in_bytes =
+        std::min(bk::device_max_mem_alloc_size(queue), bk::device_global_mem_size(queue) / 4);
+    std::int64_t part_count = 128;
+    dal::detail::check_mul_overflow(cluster_count, column_count);
+    dal::detail::check_mul_overflow(cluster_count * column_count, part_count);
+    std::int64_t fp_size = dal::detail::integral_cast<std::int64_t>(sizeof(Float));
```

Comment:
You may use `detail::get_data_type_size(detail::make_data_type<Float>())`

Reply:
Done